File: SearchIP.cpp
#include <QApplication>
#include <QWidget>
#include <QPushButton>
#include <QLineEdit>
#include <QLabel>
#include <QVBoxLayout>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QHostAddress>
#include <QStringList>
#include <QRegularExpression>

bool ipInCidr(const QString& ip, const QString& cidr) {
    // Parse CIDR, e.g., "192.168.1.0/24"
    QRegularExpression re(R"(^([\d\.]+)/(\d+)$)");
    auto match = re.match(cidr);
    if (!match.hasMatch()) return false;

    QHostAddress ip_addr(ip);
    QHostAddress network_addr(match.captured(1));
    int prefix = match.captured(2).toInt();

    if (ip_addr.protocol() != QAbstractSocket::IPv4Protocol || 
        network_addr.protocol() != QAbstractSocket::IPv4Protocol)
        return false;

    quint32 ip_int = ip_addr.toIPv4Address();
    quint32 net_int = network_addr.toIPv4Address();
    quint32 mask = prefix == 0 ? 0 : (~0U << (32 - prefix));

    return (ip_int & mask) == (net_int & mask);
}

class IPChecker : public QWidget {
    Q_OBJECT
public:
    IPChecker(QWidget* parent = nullptr) : QWidget(parent), manager(new QNetworkAccessManager(this)) {
        setWindowTitle("IP in Subnet Checker");

        ipEdit = new QLineEdit(this);
        ipEdit->setPlaceholderText("Enter IP address");

        checkBtn = new QPushButton("Check", this);
        resultLbl = new QLabel("", this);

        auto *layout = new QVBoxLayout(this);
        layout->addWidget(ipEdit);
        layout->addWidget(checkBtn);
        layout->addWidget(resultLbl);

        connect(checkBtn, &QPushButton::clicked, this, &IPChecker::onCheckClicked);
    }

private slots:
    void onCheckClicked() {
        resultLbl->setText("Checking...");
        QString ip = ipEdit->text().trimmed();
        // You may want to validate the IP format here

        // URL to your raw subnet file
        QString url = "https://raw.githubusercontent.com/CompEtcSteve/BlockList/master/ips.txt";
        QNetworkRequest req(QUrl(url));
        QNetworkReply* reply = manager->get(req);

        connect(reply, &QNetworkReply::finished, this, [this, reply, ip]() {
            if (reply->error() != QNetworkReply::NoError) {
                resultLbl->setText("Failed to fetch subnet list.");
                reply->deleteLater();
                return;
            }
            QStringList lines = QString(reply->readAll()).split('\n', Qt::SkipEmptyParts);
            bool found = false;
            for (const QString& subnet : lines) {
                QString s = subnet.trimmed();
                if (s.isEmpty()) continue;
                if (ipInCidr(ip, s)) {
                    resultLbl->setText(ip + " is in subnet " + s);
                    found = true;
                    break;
                }
            }
            if (!found)
                resultLbl->setText(ip + " is not in any of the subnets.");
            reply->deleteLater();
        });
    }

private:
    QLineEdit *ipEdit;
    QPushButton *checkBtn;
    QLabel *resultLbl;
    QNetworkAccessManager *manager;
};

#include "main.moc"

int main(int argc, char *argv[]) {
    QApplication app(argc, argv);
    IPChecker w;
    w.show();
    return app.exec();
}

File: search_IP.py
import ipaddress
import requests

def search_ip_in_subnets(ip, github_raw_url):
    ip_obj = ipaddress.ip_address(ip)

    # Fetch the file containing subnets from GitHub
    try:
        response = requests.get(github_raw_url)
        response.raise_for_status()  # Raise an error for bad HTTP responses
        subnets = response.text.splitlines()  # Split the file content into lines
    except requests.RequestException as e:
        print(f"Error fetching the file: {e}")
        return False

    # Search through the subnets
    for subnet in subnets:
        subnet = subnet.strip()  # Remove any extra whitespace/newlines
        try:
            network = ipaddress.ip_network(subnet, strict=False)
            if ip_obj in network:
                print(f"{ip} is in subnet {subnet}")
                return True
        except ValueError:
            #print(f"Invalid subnet format: {subnet}")
            continue

    print(f"{ip} is not in any of the subnets.")
    return False

# Main script to prompt for an IP and search for it in subnets
if __name__ == "__main__":
    ip_to_check = input("Enter an IP address to check: ")
    # Replace the URL below with the raw URL of your file on GitHub
    github_raw_url = 'https://raw.githubusercontent.com/CompEtcSteve/BlockList/master/ips.txt'
    search_ip_in_subnets(ip_to_check, github_raw_url)

    # Pause to allow viewing of output
    input("Press Enter to exit...")